capture_service/device_mgr.cc

```diff
@@ -47,6 +47,146 @@ std::string DeviceInfo::GetDisplayName() const
+        }
+        validated_settings.loop_single_frame_count = parsed_frame_count;
+        split_args.erase(it, it + 2);
+    }
+    if (auto it = std::find(split_args.begin(), split_args.end(), "--enable-gpu-time");
+        it != split_args.end())
+    {
+        if (settings.run_type != GfxrReplayOptions::kNormal)
+        {
+            return absl::InvalidArgumentError(
+            "Do not specify run_type = kGpuTiming in GfxrReplaySettings and also as flag "
+            "--enable-gpu-time");
+        }
+        validated_settings.run_type = GfxrReplayOptions::kGpuTiming;
+        split_args.erase(it);
+    }
+
+    // Check for run_type-specific settings
+    switch (validated_settings.run_type)
+    {
+    case GfxrReplayOptions::kPm4Dump:
+    {
+        if (!is_adreno_gpu)
+        {
+            return absl::UnimplementedError("Dump PM4 is only implemented for Adreno GPU");
+        }
+        if (validated_settings.loop_single_frame_count.has_value())
+        {
+            return absl::InvalidArgumentError(
+            "loop_single_frame_count is hardcoded for kPm4Dump, do not specify");
+        }
+        validated_settings.loop_single_frame_count = 2;
+        if (!validated_settings.metrics.empty())
+        {
+            return absl::InvalidArgumentError(
+            "Cannot use metrics except for kPerfCounters type run");
+        }
+        break;
+    }
+    case GfxrReplayOptions::kPerfCounters:
+    {
+        if (!is_adreno_gpu)
+        {
+            return absl::UnimplementedError(
+            "Perf counters feature is only implemented for Adreno GPU");
+        }
+        if (validated_settings.loop_single_frame_count.has_value())
+        {
+            return absl::InvalidArgumentError(
+            "loop_single_frame_count is hardcoded for kPerfCounters, do not specify");
+        }
+        validated_settings.loop_single_frame_count = 0;
+        if (validated_settings.metrics.size() == 0)
+        {
+            return absl::InvalidArgumentError("Must provide metrics for kPerfCounters type run");
+        }
+        break;
+    }
+    case GfxrReplayOptions::kGpuTiming:
+    default:
+    {
+        if (!validated_settings.metrics.empty())
+        {
+            return absl::InvalidArgumentError(
+            "Cannot use metrics except for kPerfCounters type run");
+        }
+        if (validated_settings.loop_single_frame_count.value_or(1) == 0)
+        {
+            return absl::InvalidArgumentError(
+            "loop_single_frame_count cannot be set to 0 for kGpuTiming and kNormal runs");
+        }
+        break;
+    }
+    }
+
+    // Re-concatenate flags to form a validated replay_flags_str
+    if (validated_settings.loop_single_frame_count.has_value())
+    {
+        assert(validated_settings.loop_single_frame_count.value() >= 0);
+        split_args.push_back("--loop-single-frame-count");
+        split_args.push_back(std::to_string(validated_settings.loop_single_frame_count.value()));
```

Comment:
nit: `validated_settings.loop_single_frame_count.value()` performs some like a `has_value()` check in order to throw the exception, which is extra work. Since you know `validated_settings.loop_single_frame_count.has_value()` you can use `*validated_settings.loop_single_frame_count` to access the value more efficiently.
```suggestion
        assert(*validated_settings.loop_single_frame_count >= 0);
        split_args.push_back("--loop-single-frame-count");
        split_args.push_back(std::to_string(*validated_settings.loop_single_frame_count));
```